### DISCUSSION TITLE
Sequel hangs when using read-only option and a transaction

### DISCUSSION THREAD

**Discussion by vitor-mda** [jeremyevans/sequel#2092] [General]
# Complete description of the issue

When having a sharded connection pool with a `read_only` option, once a transaction is opened if a statement that locks the table is executed and on a subsequent query a select operation on the locked table is done then Sequel will hang due to the select query waiting for the locked table to be released.

## Simplest Possible Self-Contained Example Showing the Issue

```ruby
DB = Sequel.postgres

DB.create_table(:xablau)

DB.transaction do
  DB.alter_table(:xablau) { add_column :column, String }

  DB[:xablau].first
end
```

## SQL Log (if any)

```
I, [2023-11-08T12:11:55.985693 #72]  INFO -- : (0.000248s) BEGIN
I, [2023-11-08T12:11:55.992592 #72]  INFO -- : (0.006680s) ALTER TABLE "xablau" ADD COLUMN "column" text
I, [2023-11-08T12:11:55.999918 #72]  INFO -- : (0.000051s) SET standard_conforming_strings = ON
I, [2023-11-08T12:11:56.000007 #72]  INFO -- : (0.000053s) SET client_min_messages = 'WARNING'
I, [2023-11-08T12:11:56.000064 #72]  INFO -- : (0.000041s) SET DateStyle = 'ISO'
I, [2023-11-08T12:11:56.000971 #72]  INFO -- : (0.000847s) SELECT CAST(current_setting('server_version_num') AS integer) AS v
```

## Expected Behavior

An error be raised instead of getting stuck waiting indefinitely for the lock to be released.

## Ruby Version
3.2.0

## Sequel Version
5.74.0

**Comment by jeremyevans** [comment]
Sequel is never going to preemptively raise an error instead of waiting for locks to be released.  If you want to avoid waiting indefinitely, then you might want to look into https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-STATEMENT-TIMEOUT

**Comment by vitor-mda** [comment]
Got it! Can the statement timeout be set via Sequel natively when creating the connection?

**Reply by jeremyevans** [reply]
Yes, use the :connect_sqls Database option.

**Reply by vitor-mda** [reply]
Awesome, that worked perfectly. Thank you very much!
